src/app/settings/admin/presets/dialog/steps/settings/provider/openstack/component.ts

```diff
@@ -87,17 +82,49 @@ export class OpenstackSettingsComponent extends BaseFormValidator implements OnI
   }
 
   private _update(): void {
-    this._presetDialogService.preset.spec.openstack = {
-      username: this.form.get(Controls.Username).value,
-      password: this.form.get(Controls.Password).value,
-      project: this.form.get(Controls.Project).value,
-      projectID: this.form.get(Controls.ProjectID).value,
+    const preset = {
       domain: this.form.get(Controls.Domain).value,
       network: this.form.get(Controls.Network).value,
       securityGroups: this.form.get(Controls.SecurityGroups).value,
       floatingIPPool: this.form.get(Controls.FloatingIPPool).value,
       routerID: this.form.get(Controls.RouterID).value,
       subnetID: this.form.get(Controls.SubnetID).value,
     } as OpenstackPresetSpec;
+
+    const credentialsType = this._determineCredentials();
+
+    // User can only specify credentials for a single credential type i.e. either default or application
+    if (credentialsType === CredentialsType.Default) {
+      preset.username = this.form.get(Controls.Credentials).value.credentials.username;
+      preset.password = this.form.get(Controls.Credentials).value.credentials.password;
+      preset.project = this.form.get(Controls.Credentials).value.credentials.project;
+      preset.projectID = this.form.get(Controls.Credentials).value.credentials.projectID;
+    } else if (credentialsType === CredentialsType.Application) {
+      preset.applicationCredentialID = this.form.get(Controls.Credentials).value.credentials.applicationCredentialID;
+      preset.applicationCredentialSecret = this.form.get(
+        Controls.Credentials
+      ).value.credentials.applicationCredentialSecret;
+    }
+    this._presetDialogService.preset.spec.openstack = preset;
+  }
+
+  private _determineCredentials(): CredentialsType {
+    if (!this.form.get(Controls.Credentials).value || !this.form.get(Controls.Credentials).value.credentials) {
+      return null;
+    }
+
+    return this._hasApplicationCredentials(this.form.get(Controls.Credentials).value.credentials)
+      ? CredentialsType.Application
+      : this._hasDefaultCredentials(this.form.get(Controls.Credentials).value.credentials)
+      ? CredentialsType.Default
+      : null;
```

Comment:
Often it's nice to use shorter form to make the code easier to read but this time I feel like this is too much. I think we should avoid using nested inline if statements as they are hard to read especially without brackets.